src/net_processing.cpp

```diff
@@ -2568,8 +2644,13 @@ bool ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStream& vR
                 const auto current_time = GetTime<std::chrono::microseconds>();
 
                 for (const CTxIn& txin : tx.vin) {
+                    // Here, we only have the txid (and not wtxid) of the
```

Comment:
I don't understand how this is supposed to work. You add the parent's txid to the node's `m_tx_process_time` via `RequestTx()` here, but when we process it in `SendMessages()` getdata logic, we always construct a `CInv` with type `MSG_WTX`. The receiving node will fail to find the tx because it searches in the mempool for a tx with that wtxid.

Reply:
This also happens if a buggy node sends `CInv` with `MSG_WTX` without having announced `wtxidrelay` first, or `MSG_TX` after having announced `wtxidrelay` I think.

Reply:
This logic of requesting unconfirmed parent transactions when an orphan is received should eventually go away, in favor of some sort of package relay scheme, where orphan transactions trigger discovery of all unconfirmed ancestors -- so I think the question is what to do in the meantime.  I think wtxid-based relay is more important than optimizing this behavior when we encounter an orphan.

Note that we only just recently (#16851, merged last October) fixed things so that these requests worked for transactions that are not in mapRelay.  Also note that for transactions that are still in mapRelay, or for non-segwit transactions where txid and wtxid are the same, these requests will still work.  However this is admittedly a regression in behavior, undoing the benefit of #16851.  However given how long that behavior was broken (and that the recent fix was largely at my urging, I believe), it's hard for me to think that anyone is really too concerned with this.

Anyway it is possible to improve this by keeping extra bookkeeping so that requests generated this way go out as MSG_TX, but I don't think it's worth it -- it's a lot of additional complexity for very little benefit, and I think effort spent on improving orphan transaction relay should be devoted to a p2p protocol extension that supports it directly.

Reply:
I disagree with the argument "this was broken for a long time and we only recently fixed it, so it's ok to break it again"

> it is possible to improve this by keeping extra bookkeeping ... it's a lot of additional complexity.

This doesn't seem like too much additional complexity. We store in `m_tx_process_time` whether the request should be for a txid or a wtxid, and set the GETDATA INV based on that. That's one additional bit of information, which could be a bool or a `enum GetDataMsg`.

> for transactions that are still in mapRelay [...] these requests will still work

Oh, I hadn't realised that we'd respond to a MSG_WTX GETDATA request for a wtxid with a transaction that has that txid. The draft BIP doesn't specify that we should do that, and it seems extremely unintuitive. Could we have two `mapRelay`s - one based on txid and one based on wtxid so we don't do this?

(I also think that we should remove mapRelay entirely (#17303), which depends on responding to NOTFOUNDs (#18238), which currently is broken and seems to be waiting for this PR (https://github.com/bitcoin/bitcoin/pull/18238#discussion_r398318859))

Reply:
>I disagree with the argument "this was broken for a long time and we only recently fixed it, so it's ok to break it again"

Please, that is not the entirety of the argument; the substance of what is being "broken" -- and how significant it is -- is relevant too.  It's not like transaction relay was horribly broken until the last release.  

For context: fixing the behavior as we did was something that I suggested we do in order to make another improvement I wanted to make -- package relay -- something that we could introduce in two parts, first as a pure client-side behavior integrated with orphan handling, and then as a p2p protocol extension.  Since then, I concluded that splitting those two things up doesn't make sense, as getting the orphan handling part right is complicated enough that it isn't worth the review burden for something that would just be ripped out later (and so I think we just go straight to package relay with a p2p protocol extension).

I think it'd be perfectly reasonable to just rip out this behavior of trying to request parents when we receive an orphan -- it is fundamentally unreliable and is already just an opportunistic way to improve relay in one edge case situation.  Adding complexity solely for the purpose of supporting something that flaky is something we should all be skeptical about; this part of the codebase is already very difficult for most people to understand.

> Oh, I hadn't realised that we'd respond to a MSG_WTX GETDATA request for a wtxid with a transaction that has that txid. The draft BIP doesn't specify that we should do that, and it seems extremely unintuitive. Could we have two mapRelays - one based on txid and one based on wtxid so we don't do this?

I don't see the downside -- the behavior I've implemented seems strictly better, though non-essential and not something I think we need to specify in a BIP as behavior we promise to maintain.

Reply:
> I don't see the downside

Just my opinion, but: the p2p code is already really complicated and not really well understood, adding new behaviours and having the implementation and the spec differ in ways that matter in practice and aren't obvious seems like it makes that a fair bit worse. This makes the review burden for both this PR and future p2p changes much higher, since you can't just read the BIPs to understand how things should work, or the local bits of code to see how they do work to see if a change makes sense, but you have to deeply understand how many things interact to even get started.

I was going to suggest adding code to explicitly ignore INVs that come in with the wrong TX/WTX marker for the peer sending them, and to not request orphan txs' parents when we're in WTX mode -- I do agree that behaviour's not too crucial to keep even without package relay; I just don't think mixing wtxid/txid values and having functionality relying on weird undocumented special cases like that working sometimes (only while the tx is in mapRelay but not after expiry?) is okay.

(Not a nack, just my opinion; I'm open to being convinced otherwise. But I think I'll go back to poking at the notfound pr rather than holding off for this one getting merged for now)

Reply:
@ajtowns My view on this is that what goes into a BIP are the essential behaviors that all implementations should abide by in order to make an idea work, but non-essential implementation details are left for projects to individually figure out.

This makes it easier for new protocol extensions to get adopted by the whole community (the fewer requirements, the better), and for our own software to stay compatible with the BIPs we claim to implement, as otherwise we'd have to constantly deprecate/redraft/repropose, as we make minor changes to our own implementation.

I'd say that how we handle a getdata request for a value that we never announced is one of those non-essential behaviors -- ignoring it is certainly fair, and returning something useful is obviously better still.  No one should assume, or would assume, that this is something that *ought* to work, so I don't think any software would be relying on this.

Also, if we had a set of design docs for how our own project implements particular BIPs, I'd agree that those should reflect more exactly what our code does -- but that would be a new initiative for us.

Reply:
I find this behavior very confusing. Can you help me understand if I'm getting any of this wrong:

1. We have a peer that has negotiate wtxid relay
2. That peer relays us a transaction whose parents we don't have
3. We want to request the parent transaction, so we put the **txid** in the `m_tx_process_time` and `m_tx_announced` for this peer (https://github.com/bitcoin/bitcoin/pull/18044/files#diff-eff7adeaec73a769788bb78858815c91R2654). That **txid** will later be added to the global `g_already_asked_for` (https://github.com/bitcoin/bitcoin/pull/18044/files#diff-eff7adeaec73a769788bb78858815c91R4151)
4. In `SendMessages()` we request that tranaction with a MSG_WTX GETDATA, but with the hash set to the **txid** (https://github.com/bitcoin/bitcoin/pull/18044/files#diff-eff7adeaec73a769788bb78858815c91R3990). 
5. The peer responds to that GETDATA with a TX because it looks in its mapRelay for transactions with a txid OR wtxid that match (https://github.com/bitcoin/bitcoin/pull/18044/files#diff-eff7adeaec73a769788bb78858815c91R1628)
6. When we receive the TX, we try to remove the **wtxid** from the peer's `m_tx_announced` and `m_tx_in_flight`, and the global `g_already_asked_for` (https://github.com/bitcoin/bitcoin/pull/18044/files#diff-eff7adeaec73a769788bb78858815c91R2587-R2594)

The thing I don't understand is that we seem to be adding the txid to these per-peer and global structures, and then trying to remove the wtxid from them. Will the txid ever be cleared? Do we end up in a request loop with the peer?

Reply:
I also can't quite follow the logic here: https://github.com/bitcoin/bitcoin/pull/18044/files#diff-eff7adeaec73a769788bb78858815c91R4139-R4140

It looks to me like the **txid** has been added to `m_tx_process_time`, but when we call into `AlreadyHave(inv, m_mempool)`, the inv is a `MSG_WTX`, so we'll check in our mempool for a transaction with that **wtxid** (which we won't find). We'll then rerequest the transaction from our peer.

Reply:
@sdaftuar "No one should assume, or would assume, that this is something that ought to work" -- but... aren't we assuming that it will work when we send a WTX request with a txid instead of a wtxid?

Aside from any bugs, I think this adds three lots of technical debt:

 * we're replying to `GETDATA WTX <txid>` and `GETDATA TX <wtxid>` if the tx is still in mapRelay -- this will go away if/when mapRelay goes away
 * we're asking for `WTX <txid>` when looking for parents of orphan transactions -- if we figure out how to replace orphan handling with package relay, this will go away too; but in the meantime it means mapRelay going away breaks orphan handling, and I think it breaks orphan handling when the parent's >15m old immediately
 * in the functional tests, as of #18446 at least, we're still sending TX INVs/GETDATAs etc despite claiming to be of a version that supports wtxid relay

One alternative might be to embrace the ambiguity: make the BIP be "if you signal wtxidrelay, then we'll send `INV TX wtxid` instead of `INV TX txid`, and also allow you to request txs by either `GETDATA TX txid` or `GETDATA TX wtxid`" ?

Reply:
I think I'm just going to eliminate the logic for requesting orphan parents from wtxid-relay peers,a as you suggested above -- I think that is the simplest approach, and should eliminate the bugs reported by @jnewbery.

Reply:
Seems okay. I think it could make sense to explicitly skip INVs offering a TX/WTX that we'd incorrectly GETDATA as a WTX/TX (respectively) as well:

```diff
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -2323,6 +2323,13 @@ bool ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStream& vR
             if (interruptMsgProc)
                 return true;
 
+            // ignore INVs that don't match wtxidrelay setting
+            if (State(pfrom->GetId())->m_wtxid_relay) {
+                if (inv.type == MSG_TX) continue;
+            } else {
+                if (inv.type == MSG_WTX) continue;
+            }
+
             bool fAlreadyHave = AlreadyHave(inv, mempool);
             LogPrint(BCLog::NET, "got inv: %s  %s peer=%d\n", inv.ToString(), fAlreadyHave ? "have" : "new", pfrom->GetId());
```

[I had a go](https://github.com/bitcoin/bitcoin/pull/18446#pullrequestreview-386416835) at updating fjahr's tests, and it seems to make reasonable sense (either with or without the change I suggest here)

(This still leaves us with the tech debt of our functional tests mostly testing tx relay when we'll be moving towards doing wtxid relay most of the time in the real world, but that's not really avoidable)

Reply:
> (This still leaves us with the tech debt of our functional tests mostly testing tx relay when we'll be moving towards doing wtxid relay most of the time in the real world, but that's not really avoidable)

I think that should be addressed with this update: https://github.com/fjahr/bitcoin/commit/becdb68ab37fdcbc1dcf743b19ea8629f52677bb

Reply:
> I think it could make sense to explicitly skip INVs offering a TX/WTX that we'd incorrectly GETDATA as a WTX/TX (respectively) as well:

And I agree with @ajtowns , this behavior makes intuitive sense to me, as it matches my interpretation of the current description in the BIP: "After a node has sent and received a "wtxidrelay" message to/from a given peer, the node is required to use the MSG_WTX inv-type when announcing transactions to that peer, or requesting announced transactions from that peer." 

I think it would even be reasonable to argue about banning peers for this behavior if there is an agreement that this behavior is wrong. The node is unnecessarily spamming the network and just silently dropping the message might let it take longer until an issue with this is discovered by the node user.